File: programmers/printer.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        
        Queue<Integer> q = new LinkedList<>();
        Queue<Integer> loc = new LinkedList<>();
        
        
        for(int i = 0 ; i < priorities.length ; i++){
            q.add(priorities[i]);
            loc.add(i);
        }
        
        int count = 0;
        
        while(true){
            int temp = q.poll();
            int index = loc.poll();
            
            if(q.isEmpty()){
                answer = count+1;
                break;
            }
            
            if( Collections.max(q) > temp ){
                q.offer(temp);
                loc.offer(index);
            }else{
                count++; // 없다면 항상 증가 해야함
                if(index == location){
                    answer = count;
                    break;
                }
                
            }
            
        }
    
        return answer;
    }
}

File: programmers/printer.py
# yaewon version
import heapq
from collections import deque

def solution(priorities, location):
    answer = 0
    
    heap = []
    queue = deque()
    for i, num in enumerate(priorities):
        heapq.heappush(heap, (-num, num))
        queue.append(i)
    
    idx = queue.popleft()
    p = heapq.heappop(heap)[1]
    
    while(1):
        if(p == priorities[idx]):
            answer +=1
            if(idx == location):
                break
            else:
                idx = queue.popleft()
                p = heapq.heappop(heap)[1]    
        else:
            queue.append(idx)
            idx = queue.popleft()
        
    return answer


# answer version - only using heapq
from collections import deque

def solution(priorities, location):
    answer = 0
    
    queue = deque([(v,i) for i,v in enumerate(priorities)])
    
    while(1):
        item = queue.popleft()
        
        # popleft로 인해서 queue가 비면 max가 안돌아가니까, if문에 queue 존재 여부도 확인 해야한다.
        if (queue and max(queue)[0] > item[0]):
            queue.append(item)
        else:
            answer += 1
            if(item[1] == location):
                break
            
        
    return answer